### PR TITLE
FIX: ProcessBuiltIn + SetBurnRole

### PR DESCRIPTION
**Pull request by mariusmihaic**
## Reasoning behind the pull request
- Fix merge conflict issue for `ProcessBuiltInFunction` and fix for `SetBurnRole`
  
## Proposed changes
- Mostly changes should be compared with this initial fix: https://github.com/multiversx/mx-chain-go/pull/6135/commits/65614615fad5d357dcffc40da8f1f732f4223045
- On top of that, added proper factories for `vmContext`

## Testing procedure
- 
- 
- 

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: vm/systemSmartContracts/sovereignEEI.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+		return nil, vm.ErrNilSystemEnvironmentInterface
+	}
+
+	return &sovereignVMContext{
+		vmContext,
+	}, nil
+}
+
+// SendGlobalSettingToAll handles sending global settings information
+func (sovHost *sovereignVMContext) SendGlobalSettingToAll(sender []byte, input []byte) error {
```
Comment (sasurobert) [inline vm/systemSmartContracts/sovereignEEI.go:27]: There is no need for this. Please revert. like really really no need.

Comment (sasurobert) [reply vm/systemSmartContracts/sovereignEEI.go:27]: remove the context handler.

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: vm/systemSmartContracts/sovereignEEI.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+	"github.com/multiversx/mx-chain-core-go/core/check"
+	"github.com/multiversx/mx-chain-go/vm"
+)
+
+type sovereignVMContext struct {
+	vm.ContextHandler
+}
+
+// NewSovereignVMContext creates a new sovereign vm context
+func NewSovereignVMContext(vmContext vm.ContextHandler) (*sovereignVMContext, error) {
```
Comment (sasurobert) [inline vm/systemSmartContracts/sovereignEEI.go:16]: This is NewSystemVMEEI - SovereignVMContext sounds bad.

It is ok that you pushed this context handler, but it needs a new name. It is the execution environmnent interface for the SystemVM.

File: vm/systemSmartContracts/eei.go [hunk 3]
```diff
@@ -826,17 +826,25 @@ func (host *vmContext) AddCode(address []byte, code []byte) {
...
 		destAcc = &vmcommon.OutputAccount{
-			Address:      scAddress,
+			Address:      input.RecipientAddr,
 			BalanceDelta: big.NewInt(0),
 		}
 		host.outputAccounts[string(destAcc.Address)] = destAcc
 	}
 
-	destAcc.BalanceDelta = big.NewInt(0).Add(destAcc.BalanceDelta, value)
+	if host.isSovereignSCToSCCall(input) {
```
Comment (sasurobert) [inline vm/systemSmartContracts/eei.go:839]: isInShardSCToSCCall is a better name.

Comment (mariusmihaic) [reply vm/systemSmartContracts/eei.go]: Right, fixed it

File: vm/systemSmartContracts/systemVMEEI.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+	"github.com/multiversx/mx-chain-core-go/core/check"
+	"github.com/multiversx/mx-chain-go/vm"
+)
+
+type systemVMEEI struct {
+	vm.ContextHandler
+}
+
+// NewSystemVMEEI creates a new system vm eei context, mainly used for now in sovereign chain
+func NewSystemVMEEI(vmContext vm.ContextHandler) (*systemVMEEI, error) {
```
Comment (sasurobert) [inline vm/systemSmartContracts/systemVMEEI.go:16]: This should be NewOneShardSystemVMEEI, the name is not good.

And should also have a rework on ProcessBuiltInFunction as well, without the sameShard IF in that case.

**Review by axenteoctavian — APPROVED** [review]

File: vm/systemSmartContracts/oneShardSystemVMEEI.go [hunk 5]
```diff
@@ -8,27 +8,28 @@ import (
...
-	return &systemVMEEI{
+	return &oneShardSystemVMEEI{
 		vmContext,
 	}, nil
 }
 
 // SendGlobalSettingToAll handles sending global settings information
-func (sovHost *systemVMEEI) SendGlobalSettingToAll(sender []byte, input []byte) error {
+func (sovHost *oneShardSystemVMEEI) SendGlobalSettingToAll(sender []byte, input []byte) error {
+	// TODO: MX-15470 remove sameShard check from ProcessBuiltInFunction
```
Comment (sasurobert) [inline vm/systemSmartContracts/oneShardSystemVMEEI.go:28]: actually have your own definition of ProcessBuiltInFunction - do not touch the ProcessBuiltInFunction from the original EEI.

**Review by sasurobert — APPROVED** [review]

**Pull request merged**